### DISCUSSION TITLE
Curso Dominando Injeção de Dependência

### DISCUSSION THREAD

**Discussion by AlexandrePatrocinio** [balta-io/forum#157] [Tira Dúvidas] [answered]
Tudo bem?

A injeção das instancias pode ser feita pelo construtor ou parâmetros dos métodos de uma classe (como no caso do .Net). No entanto também vi em minhas pesquisas que existe a técnica da injeção por propriedades em outras soluções mas é uma variação da mesma técnica de injeção de dependências.

Portanto uma dúvida: Ficou implicito, ao menos no que eu compreendi, que existe outra ou outras técnicas para resolver dependências externas de uma classe para atender o padrão de projetos IOC. Se a injeção é uma técnica de implemetação que já usa todos os meios possivéis de comunicação externa com uma instãncia de uma classe(construtor, parâmetros de métodos e propriedades), qual seria então outra técnica possível?

Não consegui imaginar outra técnica. A não ser uma que crie uma dependência direta ao service locator dentro da classe que depende de algum serviço externo.

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @AlexandrePatrocinio, como vai?

Particularmente, nunca vi injeção por propriedade... criar dependências ao meu ver é algo do construtor ou métodos.
Você tem um trecho de código que demonstre isto?

**Comment by AlexandrePatrocinio** [comment]
Vou bem e você ? 

Eu não conheço as soluções mas me parece que o Spring Framework e o PicoContainer do java usam essa abordagem (Setter Injection).   Li também que existe por interface (que responde minha pergunta) mas, ao meu ver, é como por parâmetros de um método que implementa enfim uma interface conhecida pelo service locator imagino . Artigos que eu li (inclusive o Macoratti tem artigo sobre) :

https://martinfowler.com/articles/injection.html

https://www.macoratti.net/11/07/ioc_di1.htm
